File: backend/app/schemas/job_order.py
import uuid
from decimal import Decimal

from sqlmodel import Field, SQLModel

from app.enums import JobStatus, PaymentStatus, PriceUnit
from app.models import ClaimingHistoryBase, JobItemBase, JobOrderBase, PaymentBase


class PricingData(SQLModel):
    consumption: float
    consumption_unit: PriceUnit | None = None
    rate: Decimal
    unit_price: Decimal


class JobItemExtraPublic(SQLModel):
    id: uuid.UUID
    extra_service_id: uuid.UUID
    quantity: int
    price_snapshot: Decimal
    name_snapshot: str
    
    
class JobItemExtraCreate(SQLModel):
    extra_service_id: uuid.UUID
    quantity: int
    

class JobItemPublic(JobItemBase):
    id: uuid.UUID
    total_claimed: int
    remaining_on_hand: int
    extras: list[JobItemExtraPublic] = Field(default_factory=list)
    service_id: uuid.UUID
    service_option_id: uuid.UUID
    unit_price: Decimal
    subtotal: Decimal
    service_name_snapshot: str
    service_option_name_snapshot: str
    service_abbreviation_snapshot: str
    
    
class JobItemCreate(JobItemBase):
    service_id: uuid.UUID
    service_option_id: uuid.UUID
    extras: list[JobItemExtraCreate] = Field(default_factory=list)
    

class JobItemUpdate(SQLModel):
    quantity: int | None = None
    job_status: JobStatus | None = None
    notes: str | None = None
    extra_charge: Decimal | None = None
    discount_amount: Decimal | None = None
    extras: list[JobItemExtraCreate] | None = None
    
    
class PaymentPublic(PaymentBase):
    id: uuid.UUID
    
    
class PaymentCreate(PaymentBase):
    account_id: uuid.UUID


class ClaimPublic(ClaimingHistoryBase):
    job_item_id: uuid.UUID
    
    
class ClaimCreate(ClaimingHistoryBase):
    pass


class JobOrderPublic(JobOrderBase):
    id: uuid.UUID
    job_items: list[JobItemPublic] = Field(default_factory=list)
    payments: list[PaymentPublic] = Field(default_factory=list)
    claiming_history: list[ClaimPublic] = Field(default_factory=list)
    total_due: Decimal
    total_paid: Decimal
    balance: Decimal
    customer_name: str | None = None
    customer_email: str | None = None
    customer_contact_no: str | None = None
    created_by_name: str | None = None
    updated_by_name: str | None = None
    voided_by_name: str | None = None
    void_reason: str | None = None
    

class CustomerReference(SQLModel):
    id: uuid.UUID | None = None
    name: str | None = None
    address: str | None = None
    contact_no: str | None = None
    email: str | None = None

    
class JobOrderCreate(JobOrderBase):
    override_payment_status: PaymentStatus | None = None
    customer_info: CustomerReference | None = None
    job_items: list[JobItemCreate] = Field(default_factory=list)
    payments: list[PaymentCreate] | None = None
    claiming_history: list[ClaimCreate] | None = None

File: frontend/src/types/jobOrder.ts
export type SizeUnit = 'meter' | 'in.' | 'ft.' | 'cm.' | 'mm.'
export type JobStatus =
  | 'Pending'
  | 'For Layout'
  | 'For Approval'
  | 'For Printing'
  | 'For Pickup'
  | 'Released'
  | 'Cancelled'
export type PaymentStatus = 'Fully Paid' | 'Partial' | 'Unpaid' | 'Credit' | 'Refunded' | 'Overcharged'

export interface PricingData {
  consumption: number
  consumption_unit: string
  rate: number
  unit_price: number
}

export interface Dimensions {
  height?: number
  width?: number
  size_unit?: SizeUnit
}

export interface JobItemExtra {
  id?: string
  extra_service_id: string
  quantity: number
  price_snapshot: number
  name_snapshot: string
}

export type JobItemExtraCreate = Pick<JobItemExtra, 'extra_service_id' | 'quantity'>

interface JobItemBase extends Dimensions {
  item_id: string
  description: string
  quantity: number
  job_status: JobStatus
  due_date: Date
  notes: string
  extra_charge: number
  discount_amount: number
  extras: JobItemExtra[]
}

interface ServiceRef {
  service_id: string
  service_option_id: string
}

interface ServiceSnapshot {
  service_name_snapshot: string
  service_option_name_snapshot: string
  service_abbreviation_snapshot?: string
}

interface JobItemComputed {
  unit_price: number
  subtotal: number
  total_claimed: number
  remaining_on_hand: number
}

export interface JobItemCreate extends Omit<JobItemBase, 'extras'>, ServiceRef {
  extras: JobItemExtraCreate[]
}

export type JobItemUpdate = Partial<Pick<JobItemBase, 'quantity' | 'job_status' | 'notes' | 'extra_charge' | 'discount_amount' | 'extras'>>

export interface JobItemTableRow extends JobItemBase, ServiceSnapshot, JobItemComputed {
  id?: string
}

export interface JobItem extends JobItemBase, ServiceRef, ServiceSnapshot, JobItemComputed {
  id: string
  service_abbreviation_snapshot: string
}

export interface Payment {
  date_received: Date
  reference_number: string
  amount: number
  notes: string
  account_id?: string
  account_name_snapshot: string
}

export interface ClaimingHistory {
  date_claimed: Date
  name: string
  pcs_claimed: number
  claimed_item_id: string
}

interface JobOrderBase {
  jo_number: number
  date_received: Date
  overall_job_status: JobStatus
  payment_status: PaymentStatus
}

export interface JobOrder extends JobOrderBase {
  id: string
  job_items: JobItem[]
  payments: Payment[]
  claiming_history: ClaimingHistory[]
  total_due: number
  total_paid: number
  customer_name: string
  customer_email: string
  customer_contact_no: string
  created_at: Date
  updated_at: Date
  created_by_name: string
  updated_by_name: string
}

export interface JobOrderCreate extends JobOrderBase {
  customer_id?: string
  customer_name?: string
  customer_address?: string
  customer_contact_no?: string
  customer_email?: string
  job_items: JobItemCreate[]
  payments?: Payment[]
  claims?: ClaimingHistory[]
}
